### PR TITLE
Add jth_hess_coord!, jth_hprod! and ghjvprod!

### PR DESCRIPTION
**Pull request by amontoison**
@tmigot Do we have some unit tests for these routines?

### CODE REVIEW

File: src/moi_nlp_model.jl [hunk 1]
```diff
@@ -413,3 +445,48 @@ function NLPModels.hprod!(
...
+end
+
+function NLPModels.ghjvprod!(
+  nlp::MathOptNLPModel,
+  x::AbstractVector,
+  g::AbstractVector,
+  v::AbstractVector,
+  ghv::AbstractVector,
+)
+  # Don't we have a counter :neval_ghjvprod?
```
Comment (tmigot) [inline src/moi_nlp_model.jl:477]:
```suggestion
  increment!(nlp, :neval_hprod)
```

File: src/moi_nlp_model.jl [hunk 2]
```diff
@@ -349,6 +349,38 @@ function NLPModels.hess_coord!(
...
 end
 
+function NLPModels.jth_hess_coord!(
+  nlp::MathOptNLPModel,
+  x::AbstractVector,
+  j::Integer,
+  vals::AbstractVector,
+)
+  increment!(nlp, :neval_jhess)
+  vals .= 0.0
```
Comment (tmigot) [inline src/moi_nlp_model.jl:359]:
```suggestion
  increment!(nlp, :neval_jhess)
  @rangecheck 1 nlp.meta.ncon j
  vals .= 0.0
```

File: src/moi_nlp_model.jl [hunk 3]
```diff
@@ -413,3 +445,48 @@ function NLPModels.hprod!(
...
+
+function NLPModels.jth_hprod!(
+  nlp::MathOptNLPModel,
+  x::AbstractVector,
+  v::AbstractVector,
+  j::Integer,
+  hv::AbstractVector,
+)
+  increment!(nlp, :neval_jhprod)
+  hv .= 0.0
```
Comment (tmigot) [inline src/moi_nlp_model.jl:457]:
```suggestion
  increment!(nlp, :neval_jhprod)
  @rangecheck 1 nlp.meta.ncon j
  hv .= 0.0
```

File: src/moi_nlp_model.jl [hunk 4]
```diff
@@ -413,3 +445,48 @@ function NLPModels.hprod!(
...
+  if nlp.meta.nlin + 1 ≤ j ≤ nlp.meta.nlin + nlp.quadcon.nquad
+    qcon = nlp.quadcon.constraints[j - nlp.meta.nlin]
+    coo_sym_add_mul!(qcon.A.rows, qcon.A.cols, qcon.A.vals, v, hv, 1.0)
+  end
+  if nlp.meta.nlin + nlp.quadcon.nquad + 1 ≤ j ≤ nlp.meta.ncon
```
Comment (tmigot) [inline src/moi_nlp_model.jl:462]:
```suggestion
  else
```

File: src/moi_nlp_model.jl [hunk 5]
```diff
@@ -349,6 +349,38 @@ function NLPModels.hess_coord!(
...
+    index = nlp.obj.nnzh
+    for i = 1:(nlp.quadcon.nquad)
+      qcon = nlp.quadcon.constraints[i]
+      if j == nlp.meta.nlin + i
+        view(vals, (index + 1):(index + qcon.nnzh)) .= qcon.A.vals
+      end
+      index += qcon.nnzh
+    end
+  end
+  if nlp.meta.nlin + nlp.quadcon.nquad + 1 ≤ j ≤ nlp.meta.ncon
```
Comment (tmigot) [inline src/moi_nlp_model.jl:370]:
```suggestion
  else
```

File: src/moi_nlp_model.jl [hunk 6]
```diff
@@ -349,6 +349,38 @@ function NLPModels.hess_coord!(
...
+    nlp.λ[j - nlp.meta.nlin - nlp.quadcon.nquad] = 1.0
```
Comment (tmigot) [inline src/moi_nlp_model.jl:371]: How do we know this vector is zeros?

File: src/moi_nlp_model.jl [hunk 7]
```diff
@@ -349,6 +349,38 @@ function NLPModels.hess_coord!(
...
+    MOI.eval_hessian_lagrangian(
+      nlp.eval,
+      view(vals, (nlp.obj.nnzh + nlp.quadcon.nnzh + 1):(nlp.meta.nnzh)),
+      x,
+      0.0,
+      λ,
+    )
+    nlp.λ[j - nlp.meta.nlin - nlp.quadcon.nquad] = 0.0
```
Comment (tmigot) [inline src/moi_nlp_model.jl:379]: If we do these things, maybe add a `reset!` function?

File: src/moi_nlp_model.jl [hunk 8]
```diff
@@ -413,3 +445,48 @@ function NLPModels.hprod!(
...
+  hv::AbstractVector,
+)
+  increment!(nlp, :neval_jhprod)
+  hv .= 0.0
+  if nlp.meta.nlin + 1 ≤ j ≤ nlp.meta.nlin + nlp.quadcon.nquad
+    qcon = nlp.quadcon.constraints[j - nlp.meta.nlin]
+    coo_sym_add_mul!(qcon.A.rows, qcon.A.cols, qcon.A.vals, v, hv, 1.0)
+  end
+  if nlp.meta.nlin + nlp.quadcon.nquad + 1 ≤ j ≤ nlp.meta.ncon
+    nlp.λ[j - nlp.meta.nlin - nlp.quadcon.nquad] = 1.0
```
Comment (tmigot) [inline src/moi_nlp_model.jl:463]: see above

File: src/moi_nlp_model.jl [hunk 9]
```diff
@@ -413,3 +445,48 @@ function NLPModels.hprod!(
...
+    MOI.eval_hessian_lagrangian_product(nlp.eval, hv, x, v, 0.0, nlp.λ)
+    nlp.λ[j - nlp.meta.nlin - nlp.quadcon.nquad] = 0.0
```
Comment (tmigot) [inline src/moi_nlp_model.jl:465]: see above

File: src/moi_nlp_model.jl [hunk 10]
```diff
@@ -413,3 +445,48 @@ function NLPModels.hprod!(
...
+)
+  # Don't we have a counter :neval_ghjvprod?
+  ghv .= 0.0
+  for i in nlp.meta.nlin + 1 : nlp.meta.nlin + nlp.quadcon.nquad
+    qcon = nlp.quadcon.constraints[i - nlp.meta.nlin]
+    ghv[i] = coo_sym_dot(qcon.A.rows, qcon.A.cols, qcon.A.vals, g, v)
+  end
+  if nlp.meta.nnln > 0
+    # we could store hv in the MathOptNLPModel
+    hv = zeros(Float64, nlp.meta.nvar)
```
Comment (tmigot) [inline src/moi_nlp_model.jl:485]: Please make an issue for that

**Comment by tmigot** [comment]
> @tmigot Do we have some unit tests for these routines?

You need to add `exclude = []` here https://github.com/amontoison/NLPModelsJuMP.jl/blob/10ccc77db66baad1f357f71fe22f57eea10b8350/test/nlp_consistency.jl#L7

The unit tests are https://github.com/JuliaSmoothOptimizers/NLPModelsTest.jl/blob/main/src/nlp/consistency.jl but excluded by default for these.

**Review by tmigot — CHANGES_REQUESTED** [review]
Thanks @amontoison ! just some comments

File: src/moi_nlp_model.jl (hunk 8 above)
Comment (amontoison) [reply src/moi_nlp_model.jl:463]: I commited your suggestion but it's wrong. We have a issue is `j` is an index of linear constraint of if `j` is not an index of a valid constraint.

File: src/moi_nlp_model.jl [hunk 11]
```diff
@@ -349,6 +349,38 @@ function NLPModels.hess_coord!(
...
+      index += qcon.nnzh
+    end
+  else
+    nlp.λ[j - nlp.meta.nlin - nlp.quadcon.nquad] = 1.0
+    MOI.eval_hessian_lagrangian(
+      nlp.eval,
+      view(vals, (nlp.obj.nnzh + nlp.quadcon.nnzh + 1):(nlp.meta.nnzh)),
+      x,
+      0.0,
+      λ,
```
Comment (amontoison) [inline src/moi_nlp_model.jl:377]:
```suggestion
      nlp.λ,
```

**Comment by amontoison** [comment]
@tmigot I don't understand the CI error.
It seems related to the counter for `neval_jhprod`.

File: src/moi_nlp_model.jl (hunk 10 above)
Comment (amontoison) [reply src/moi_nlp_model.jl]: I added `hv` in the structure of `MathOptNLPModel`.

File: src/moi_nlp_model.jl [hunk 12]
```diff
@@ -413,3 +448,45 @@ function NLPModels.hprod!(
...
+  ghv::AbstractVector,
+)
+  increment!(nlp, :neval_hprod)
+  ghv .= 0.0
+  for i in nlp.meta.nlin + 1 : nlp.meta.nlin + nlp.quadcon.nquad
+    qcon = nlp.quadcon.constraints[i - nlp.meta.nlin]
+    ghv[i] = coo_sym_dot(qcon.A.rows, qcon.A.cols, qcon.A.vals, g, v)
+  end
+  for i in nlp.meta.nlin + nlp.quadcon.nquad + 1 : nlp.meta.ncon
+    jth_hprod!(nlp, x, v, i, nlp.hv)
```
Comment (tmigot) [inline src/moi_nlp_model.jl:488]: This is your extra evaluations
```suggestion
    jth_hprod!(nlp, x, v, i, nlp.hv)
    decrement!(nlp, :neval_jhprod)
```

File: src/moi_nlp_model.jl (hunk 8 above)
Comment (tmigot) [reply src/moi_nlp_model.jl:467]: Argh, sorry. Should've been a elseif then?

**Review by tmigot — CHANGES_REQUESTED** [review]

File: src/moi_nlp_model.jl (hunk 7 above)
Comment (tmigot) [reply src/moi_nlp_model.jl:382]: Yeah... not necessary. Your choice :).

File: src/moi_nlp_model.jl (hunk 8 above)
Comment (amontoison) [reply src/moi_nlp_model.jl:465]: Yes, I added a commit with two `elseif`.

File: src/moi_nlp_model.jl (hunk 12 above)
Comment (tmigot) [reply src/moi_nlp_model.jl:486]: This is still the line you need to add for the counters

**Review by tmigot — APPROVED** [review]
Thanks @amontoison !

**Pull request merged**